core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/UserIdResolverCache.java

```diff
@@ -0,0 +1,226 @@
+                                    + entry + " for the key: " + key + " successfully");
+                }
+            }
+        } finally {
+            PrivilegedCarbonContext.endTenantFlow();
+        }
+    }
+
+    /**
+     * Retrieves a cache entry.
+     *
+     * @param key       CacheKey.
+     * @param cacheName Name of the cache.
+     * @return Cached entry if the key presents, else returns null.
+     */
+    public String getValueFromCache(String key, String cacheName) {
+
+        if (StringUtils.isEmpty(key) || StringUtils.isEmpty(cacheName)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Invalid input parameters in get value from cache request. Cache key: " + key +
+                        " ,Cache name: " + cacheName);
+            }
+            return null;
+        }
+
+        PrivilegedCarbonContext.startTenantFlow();
+
+        try {
+            PrivilegedCarbonContext carbonContext = PrivilegedCarbonContext.getThreadLocalCarbonContext();
+            carbonContext.setTenantId(MultitenantConstants.SUPER_TENANT_ID);
+            carbonContext.setTenantDomain(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME);
+
+            Cache<String, String> cache = UserIdResolverCache(cacheName);
+            if (cache != null) {
+                if (cache.containsKey(key)) {
+                    String entry = cache.get(key);
+                    if (log.isDebugEnabled()) {
+                        log.debug(cacheName + " which is under " + USER_ID_RESOLVER_CACHE_MANAGER +
+                                ", found the entry: " + entry + " for the key: " + key + " successfully.");
+                    }
+                    return entry;
+                }
+                if (log.isDebugEnabled()) {
+                    log.debug(cacheName + " which is under " + USER_ID_RESOLVER_CACHE_MANAGER +
+                            ", doesn't contain the key: " + key);
+                }
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("Error while getting the cache : " + cacheName + " which is under "
+                            + USER_ID_RESOLVER_CACHE_MANAGER);
+                }
+            }
+            return null;
+        } finally {
+            PrivilegedCarbonContext.endTenantFlow();
+        }
+    }
+
+    /**
+     * Clears a cache entry.
+     *
+     * @param key       Key to clear cache.
+     * @param cacheName Name of the cache.
+     */
+    public void clearCacheEntry(String key, String cacheName) {
+
+        if (StringUtils.isEmpty(key) || StringUtils.isEmpty(cacheName)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Invalid input parameters in clear from cache request. Cache key: " + key +
+                        " ,Cache name: " + cacheName);
+            }
+            return;
+        }
+
+        PrivilegedCarbonContext.startTenantFlow();
```

Comment:
Move this inside try block and extract a private method named. startSuperTenantFlow

Reply:
fixed